File: scripts/generate_audio.js
#!/usr/bin/env node

/**
 * 使用 Web Speech API 和浏览器生成音频
 * 或使用 Google Translate TTS API (免费，无需 key)
 */

const https = require("https");
const fs = require("fs");
const path = require("path");

const OUTPUT_DIR = path.join(__dirname, "../public/audio");

// 确保输出目录存在
if (!fs.existsSync(OUTPUT_DIR)) {
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

// 使用 Google Translate 的免费 TTS API
async function generateAudioFromGoogle(text, filename) {
  return new Promise((resolve, reject) => {
    try {
      // Google Translate TTS URL
      const encodedText = encodeURIComponent(text);
      const url = `https://translate.google.com/translate_tts?ie=UTF-8&client=tw-ob&q=${encodedText}&tl=en`;

      const file = fs.createWriteStream(path.join(OUTPUT_DIR, filename));

      https
        .get(
          url,
          {
            headers: {
              "User-Agent":
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
            },
          },
          (response) => {
            if (response.statusCode !== 200) {
              reject(new Error(`HTTP ${response.statusCode}`));
              return;
            }
            response.pipe(file);
            file.on("finish", () => {
              file.close();
              resolve();
            });
          },
        )
        .on("error", reject);
    } catch (error) {
      reject(error);
    }
  });
}

// 所有需要生成的文本
const textToGenerate = [
  // 成功时
  { filename: "encouragement-yes.mp3", text: "Excellent!" },
  { filename: "encouragement-thatsit.mp3", text: "Well done!" },
  { filename: "encouragement-go.mp3", text: "Perfect!" },
  { filename: "encouragement-boom.mp3", text: "Awesome!" },
  { filename: "encouragement-comeon.mp3", text: "Fantastic!" },
  { filename: "encouragement-push.mp3", text: "Amazing!" },
  { filename: "encouragement-waytogo.mp3", text: "Brilliant!" },
  { filename: "encouragement-onemore.mp3", text: "Outstanding!" },
  // 失败时
  { filename: "miss-again.mp3", text: "Try again!" },
  { filename: "miss-comeon.mp3", text: "Keep going!" },
  { filename: "miss-push.mp3", text: "You can do it!" },
  { filename: "miss-nextone.mp3", text: "Next one!" },
  { filename: "miss-go.mp3", text: "Don't give up!" },
  { filename: "miss-letsgo.mp3", text: "Come on!" },
];

async function generateAll() {
  console.log("开始生成音频文件...\n");
  let success = 0;
  let failed = 0;

  for (const item of textToGenerate) {
    try {
      console.log(`生成: ${item.filename}`);
      console.log(`文本: ${item.text}`);
      await generateAudioFromGoogle(item.text, item.filename);
      console.log("✓ 成功\n");
      success++;

      // 延迟以避免被限流
      await new Promise((resolve) => setTimeout(resolve, 1000));
    } catch (error) {
      console.log(`✗ 失败: ${error.message}\n`);
      failed++;
    }
  }

  console.log("=".repeat(50));
  console.log(
    `生成完成！ 成功: ${success}/${textToGenerate.length}, 失败: ${failed}`,
  );
  console.log("=".repeat(50));

  // 列出生成的文件
  const files = fs.readdirSync(OUTPUT_DIR);
  console.log(`\n生成的文件列表 (${files.length} 个):`);
  files.forEach((f) => {
    const filePath = path.join(OUTPUT_DIR, f);
    const stats = fs.statSync(filePath);
    console.log(`  - ${f} (${(stats.size / 1024).toFixed(1)} KB)`);
  });
}

generateAll().catch(console.error);

File: scripts/generate_audio.py
#!/usr/bin/env python3
"""
生成鼓励语音频文件脚本
使用 pyttsx3 生成高质量的人声音频
"""

import os
import sys
import pyttsx3

# 输出目录
OUTPUT_DIR = os.path.join(os.path.dirname(__file__), '../public/audio')

# 确保输出目录存在
os.makedirs(OUTPUT_DIR, exist_ok=True)

# 初始化 TTS 引擎
engine = pyttsx3.init()

# 配置语音参数 - 提高音量和音高使听起来更有活力
engine.setProperty('rate', 150)  # 稍快的语速
engine.setProperty('volume', 0.9)  # 音量 0-1

# 获取可用的语音并选择女性语音（通常听起来更有活力）
voices = engine.getProperty('voices')
print(f"可用语音: {len(voices)} 个")
for i, voice in enumerate(voices):
    print(f"  {i}: {voice.name} (ID: {voice.id})")

# 尝试选择英文语音
selected_voice = voices[1] if len(voices) > 1 else voices[0]
engine.setProperty('voice', selected_voice.id)
print(f"使用语音: {selected_voice.name}\n")

# 鼓励语列表 - 成功时
encouragements = [
    ("encouragement-yes.mp3", "Yes! Yes! Yes!"),
    ("encouragement-thatsit.mp3", "That's it! That's it!"),
    ("encouragement-go.mp3", "Go! Go! Go!"),
    ("encouragement-boom.mp3", "Boom! Boom! Boom!"),
    ("encouragement-comeon.mp3", "Come on! Come on!"),
    ("encouragement-push.mp3", "Push! Push! Push!"),
    ("encouragement-waytogo.mp3", "Way to go! Way to go!"),
    ("encouragement-onemore.mp3", "One more! One more!"),
]

# 鼓励语列表 - 失败时
miss_encouragements = [
    ("miss-again.mp3", "Again! Again! Again!"),
    ("miss-comeon.mp3", "Come on! Don't stop!"),
    ("miss-push.mp3", "Push! Push harder!"),
    ("miss-nextone.mp3", "Next one! Next one!"),
    ("miss-go.mp3", "Go! Go! Go!"),
    ("miss-letsgo.mp3", "Let's go! Let's go!"),
]

print("=" * 50)
print("生成成功时的鼓励语音频...")
print("=" * 50)

# 生成成功时的音频
for filename, text in encouragements:
    filepath = os.path.join(OUTPUT_DIR, filename)
    print(f"生成: {filename}")
    print(f"  文本: {text}")
    try:
        engine.save_to_file(text, filepath)
        engine.runAndWait()
        print(f"  ✓ 成功保存\n")
    except Exception as e:
        print(f"  ✗ 错误: {e}\n")

print("=" * 50)
print("生成失败时的鼓励语音频...")
print("=" * 50)

# 生成失败时的音频
for filename, text in miss_encouragements:
    filepath = os.path.join(OUTPUT_DIR, filename)
    print(f"生成: {filename}")
    print(f"  文本: {text}")
    try:
        engine.save_to_file(text, filepath)
        engine.runAndWait()
        print(f"  ✓ 成功保存\n")
    except Exception as e:
        print(f"  ✗ 错误: {e}\n")

print("=" * 50)
print(f"✓ 所有音频文件已生成到: {OUTPUT_DIR}")
print("=" * 50)

# 列出生成的文件
audio_files = os.listdir(OUTPUT_DIR)
print(f"\n生成的文件列表 ({len(audio_files)} 个):")
for f in sorted(audio_files):
    file_path = os.path.join(OUTPUT_DIR, f)
    file_size = os.path.getsize(file_path) / 1024  # KB
    print(f"  - {f} ({file_size:.1f} KB)")
